File: pyg4ometry/pygeometry/geant4/solid/Cons.py
from SolidBase import SolidBase as _SolidBase
from pygeometry.pycsg.core import CSG as _CSG
from pygeometry.geant4.Registry import registry as _registry
from pygeometry.geant4.solid.Plane import Plane as _Plane
from pygeometry.geant4.solid.Wedge import Wedge as _Wedge
from pygeometry.pycsg.geom import Vector as _Vector
import numpy as _np

class Cons(_SolidBase):
    """
    Constructs a conical section.

    Inputs:
        name:      string, name of the volume
        pRMin1:    float, inner radius at -pDz
        pRMax1:    float, outer radius at -pDz
        pRMin2:    float, inner radius at +pDZ
        pRMax2:    float, outer radius at +pDz
        pDz:       float, half-length along z
        pSPhi:     float, starting phi angle
        pDPhi:     float, angle of segment in radians
    """

    def __init__(self, name, pRmin1, pRmax1, pRmin2, pRmax2, pDz, pSPhi, pDPhi):
        self.name = name
        self.type = 'Cons'
        self.pRmin1 = pRmin1
        self.pRmax1 = pRmax1
        self.pRmin2 = pRmin2
        self.pRmax2 = pRmax2
        self.pDz = pDz
        self.pSPhi = pSPhi
        self.pDPhi = pDPhi
        self.mesh = None
        _registry.addSolid(self)
        self.checkParameters()

    def checkParameters(self):
        if self.pRmin1 > self.pRmax1:
            raise ValueError("Inner radius must be less than outer radius.")
        if self.pRmin2 > self.pRmax2:
            raise ValueError("Inner radius must be less than outer radius.")
        if self.pDPhi > _np.pi*2:
            raise ValueError("pDTheta must be less than 2 pi")

    def pycsgmesh(self):
        self.basicmesh()
        self.csgmesh()

        return self.mesh


    def basicmesh(self):
        if self.pRmax1 < self.pRmax2:
            self.R1 = self.pRmax2  # Cone with tip pointing towards -z
            self.r1 = self.pRmin2
            self.R2 = self.pRmax1
            self.r2 = self.pRmin1
            self.factor = -1

        else:
            self.R1 = self.pRmax1  # Cone with tip pointing towards +z
            self.r1 = self.pRmin1
            self.R2 = self.pRmax2
            self.r2 = self.pRmin2
            self.factor = 1

        h = 2 * self.pDz
        self.H1 = float(self.R2 * h) / float(self.R1 - self.R2)

        try:  # Avoid crash when both inner radii are 0
            self.H2 = float(self.r2 * h) / float(self.r1 - self.r2)
        except ZeroDivisionError:
            self.H2 = 0

        self.h1 = self.factor * (h + self.H1)
        self.h2 = self.factor * (h + self.H2)

        self.mesh = _CSG.cone(start=[0, 0, -self.factor * self.pDz], end=[0, 0, self.h1 - self.factor * self.pDz], radius=self.R1)

        return self.mesh

    def csgmesh(self):

        
        # ensure radius for intersection wedge is much bigger than solid
        wrmax = 3 * (self.pRmax1 + self.pRmax2)
        # ensure the plane is large enough to cut cone correctly. Extroplates the z intersect using two points.
        wzlength = abs((self.pRmax2*((2*self.pDz)/(self.pRmax2 - self.pRmax1))) + self.pDz)   

        if self.pDPhi != 2 * _np.pi:
            pWedge = _Wedge("wedge_temp", wrmax, self.pSPhi, self.pSPhi + self.pDPhi, wzlength).pycsgmesh()
        else:
            pWedge = _CSG.cylinder(start=[0, 0, -self.pDz * 5], end=[0, 0, self.pDz * 5],
                                   radius=self.R1 * 5)  # factor 5 is just to ensure wedge mesh is much bigger than cone solid

        pTopCut = _Plane("pTopCut_temp", _Vector(0, 0, 1), self.pDz, wzlength).pycsgmesh()
        pBotCut = _Plane("pBotCut_temp", _Vector(0, 0, -1), -self.pDz, wzlength).pycsgmesh()

        if self.H2:
            sInner = _CSG.cone(start=[0, 0, -self.factor * self.pDz], end=[0, 0, self.h2 - self.factor * self.pDz], radius=self.r1)
            self.mesh = self.mesh.subtract(sInner).intersect(pWedge).subtract(pBotCut).subtract(pTopCut)
        else:
            self.mesh = self.mesh.intersect(pWedge).subtract(pBotCut).subtract(pTopCut)

        return self.mesh

File: pyg4ometry/pygeometry/cpgeant4/solid/Cons.cpp
#include "Cons.h"

CSG* CSGMesh::ConstructCons(double pRmin1,double pRmax1,double pRmin2,double pRmax2,double pDz,double pSPhi,double pDPhi){
  double R1,r1,R2,r2;
  double factor;
  if(pRmax1 < pRmax2){
    R1 = pRmax2;
    r1 = pRmin2;
    R2 = pRmax1;
    r2 = pRmin1;
    factor = -1.0;
  }
  else{
    R1 = pRmax1;
    r1 = pRmin1;
    R2 = pRmax2;
    r2 = pRmin2;
    factor = 1.0;
  }
  double h = 2.0*pDz; 
  double H1 = (R2*h)/(R1-R2);
  double H2 = 0.;
  if(r1-r2 != 0.0){
    H2 = (r2*h)/(r1-r2);
  }

  double h1 = factor*(h+H1);
  double h2 = factor*(h+H2);

  CSG* basicmesh = Solids::Cone(new Vector(0.0,0.0,-factor*pDz),new Vector(0.0,0.0,h1-factor),R1);


  double wrmax = 3.0*(pRmax1+pRmax2); // ensure radius for intersection wedge is much bigger than solid
  double wzlength = 3.0*pDz;

  CSG* pWedge; 
  if( pDPhi != 2.0*M_PI ){
    SolidBase* wedge = new Wedge("wedge_temp",wrmax,pSPhi,pSPhi+pDPhi,wzlength);
    pWedge = wedge->GetMesh();
  }
  else{
    pWedge = Solids::Cylinder(5.*pDz,5.*R1);
  }

  G4Plane* TopCutPlane = new G4Plane("pTopCut_temp",new Vector(0.0,0.0,1.0),pDz,wzlength);
  CSG* pTopCut = TopCutPlane->GetMesh();

  G4Plane* BotCutPlane = new G4Plane("pBotCut_temp",new Vector(0.0,0.0,-1.0),-pDz,wzlength);
  CSG* pBotCut = BotCutPlane->GetMesh();

  CSG* mesh;
  if(H2 != 0.){
    CSG* sInner = Solids::Cone(new Vector(0.,0.,-factor*pDz),new Vector(0.0,0.0,h2-factor*pDz),r1);
    mesh = basicmesh->Subtract(sInner)->Intersect(pWedge)->Subtract(pBotCut)->Subtract(pTopCut);
  }
  else{
    mesh = basicmesh->Intersect(pWedge)->Subtract(pBotCut)->Subtract(pTopCut); 
  }
  return mesh;
}
